File: src/p543/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        result = 0

        def DFS(root):
            if root is None:
                return 0
            l = DFS(root.left)
            r = DFS(root.right)
            ll = 0 if root.left is None else root.left.l
            lr = 0 if root.right is None else root.right.l
            root.l = max(ll, lr) + 1
            return max(l, r, ll + lr)

        return DFS(root)

File: src/p543/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  typedef pair<int, int> pi;
  pi run(TreeNode* root) {
    if (!root) return {0, 0};
    pi result, left, right;
    if (root->left) {
      left = run(root->left);
    }
    if (root->right) {
      right = run(root->right);
    }
    result.first = max(left.first, right.first) + 1;
    result.second = left.first + right.first;
    int child = max(left.second, right.second);
    result.second = max(result.second, child);
    return result;
  }
  int diameterOfBinaryTree(TreeNode* root) {
    return run(root).second;
  }
};
